### PR TITLE
Workflow engine refactor

### PR DESCRIPTION
**Pull request by benbp**
Refactoring the workflow engine for:

- High Availability
- Fault Tolerance
- Better performance
- Better execution strategy

This PR includes mostly minor refactoring to expose the task library as an injectable, slight changes in how Task.run() is used, and updates to the run-graph job based on the refactor of TaskGraph to on-core.

Relies on https://github.com/RackHD/on-taskgraph/pull/56, https://github.com/RackHD/on-core/pull/75, and https://github.com/RackHD/on-http/pull/112

See the description of https://github.com/RackHD/on-taskgraph/pull/56 for architectural slides.

### CODE REVIEW

**Comment by benbp** [comment]
Mentioning a mix of those who came to the presentation/should review this: @RackHD/corecommitters @VulpesArtificem @heckj @king-jam @johren @stuart-stanley @jimturnquist @larry-dean @davequick @zyoung51 @zhangy42 @mattgam @tldavies @DavidjohnBlodgett @rolandpoulter

**Comment by benbp** [comment]
@rolandpoulter yeah, the ideal way I think to keep track of this is to add a start time to the taskdependency document in the database when it is checked out for the first time (i.e. don't reset the timer if a task-runner fails and another one picks up the task), in which case doing that query is quite trivial. Then we can just poll the database for task timeouts periodically and queue up cancellation events. We just ended up not addressing it in this first rev and were going to improve it as follow-on work unless anyone thinks it's important enough to block on.

**Comment by rolandpoulter** [comment]
:+1:

File: lib/jobs/base-job.js [hunk 1]
```diff
@@ -128,25 +127,20 @@ function baseJobFactory(
         if (error) {
             loggerObject.error = error;
         }
-        this.logger.info("Cancelling job.", loggerObject);
-        this._done(error);
+        this.logger.info("Stopping job.", loggerObject);
+        return this._done(error);
```
Comment (tldavies) [inline lib/jobs/base-job.js]: newline

File: lib/jobs/run-graph.js [hunk 2]
```diff
@@ -52,34 +69,39 @@ function runGraphJobFactory(BaseJob, taskGraphProtocol, Logger, assert, uuid, ut
...
-                // https://github.com/RackHD/docs/pull/222.
-                logger.error("Error starting graph for task.", {
-                    taskId: self.taskId,
-                    graphName: self.options.graphName,
-                    error: e
-                });
-                self._done(new Error("Unable to start graph for task " + self.taskId +
-                                     " in graph " + self.options.graphName));
+        taskGraphStore.findActiveGraphForTarget(self.graphTarget)
+        .then(function(activeGraph) {
```
Comment (tldavies) [inline lib/jobs/run-graph.js]: indentation is off

File: lib/task.js [hunk 3]
```diff
@@ -238,141 +209,100 @@ function factory(
...
-            this.state = TaskStates.Failed;
-            this.error = e;
-            return this.finish();
-        }
-
-        return this._run();
-    };
-
-    Task.prototype.finish = function() {
+    Task.prototype._run = function() {
```
Comment (tldavies) [inline lib/task.js]: Difference between _run and run

File: lib/jobs/install-os.js [hunk 4]
```diff
@@ -65,7 +65,6 @@ function installOsJobFactory(
      */
     InstallOsJob.prototype._validateOptions = function() {
         assert.string(this.context.target);
-        assert.string(this.options.completionUri);
```
Comment (heckj) [inline lib/jobs/install-os.js]: intentional removal?

Comment (benbp) [inline lib/jobs/install-os.js]: Hmm, I'll check, it's possible I need to rebase this out, but I do believe it does not exist in the task definition anymore.

File: lib/task.js (hunk 3 above)
Comment (benbp) [inline lib/task.js]: Outer `run` calls inner `_run` with a bunch of error handling conditionals and so forth. It was more readable to split these out.

**Comment by heckj** [comment]
:+1:

File: lib/task.js (hunk 3 above)
Comment (tldavies) [inline lib/task.js]: ok, can you add that to the comments. Cause that is not obvious.

**Comment by jlongever** [comment]
+1 assuming those check failures are due to dependent PR's.

**Pull request merged**